File: 백준/Gold/2623. 음악프로그램/음악프로그램.js
const fs = require('fs');
const input = fs.readFileSync(0, 'utf-8').toString().trim().split('\n');
const [[N, M], ...arr] = input.map(v => v.split(' ').map(Number));

const degree = Array(N+1).fill(0);
const graph = Array.from({length:N+1}, () => []);
for (const order of arr){
    for (let i=1; i<order[0]; i++){
        degree[order[i+1]]++;
        graph[order[i]].push(order[i+1]);
    }
}

const ans = [];
const queue = [];

for (let i=1; i<N+1; i++){
    if(degree[i] == 0) queue.push(i);
}

while (queue.length > 0){
    const now = queue.shift();
    ans.push(now);
    for (const next of graph[now]){
        degree[next]--;
        if(degree[next] == 0) queue.push(next);
    }
}
if(ans.length != N) console.log(0);
else console.log(ans.join('\n'));

File: 백준/Gold/2623. 음악프로그램/음악프로그램.py
import sys
from collections import deque
input = sys.stdin.readline

N, M = map(int, input().split())
indegree = [0]*(N+1)
graph = [[] for _ in range(N+1)]

for _ in range(M):
    order = list(map(int, input().split()))
    for i in range(1, order[0]):
        graph[order[i]].append(order[i+1])
        indegree[order[i+1]] += 1

queue = deque()
for i in range(1,N+1):
    if indegree[i] == 0:
        queue.append(i)

answer = []
while queue:
    now = queue.popleft()
    answer.append(now)
    
    for next in graph[now]:
        indegree[next] -= 1
        if indegree[next] == 0:
            queue.append(next)

if len(answer) == N:
    print('\n'.join(map(str, answer)))
else :
    print(0)
